components/fxa-client/src/internal/close_remote_tabs.rs

```diff
@@ -0,0 +1,98 @@
+
+use super::{
+    commands::{
+        close_remote_tabs::{self, CloseTabsPayload, EncryptedCloseTabsPayload},
+        send_tab::PrivateSendTabKeys,
+        IncomingDeviceCommand,
+    },
+    http_client::GetDeviceResponse,
+    scopes, FirefoxAccount,
+};
+use crate::{Error, Result};
+
+impl FirefoxAccount {
+    pub fn close_remote_tabs<T: AsRef<str>>(
+        &mut self,
+        target_device_id: &str,
+        urls: &[T],
+    ) -> Result<()> {
+        let devices = self.get_devices(false)?;
+        let target = devices
+            .iter()
+            .find(|d| d.id == target_device_id)
+            .ok_or_else(|| Error::UnknownTargetDevice(target_device_id.to_owned()))?;
+        let payload = CloseTabsPayload {
+            urls: urls.iter().map(|url| url.as_ref().to_owned()).collect(),
+        };
+        let oldsync_key = self.get_scoped_key(scopes::OLD_SYNC)?;
+        let command_payload =
+            close_remote_tabs::build_close_remote_tabs_command(oldsync_key, target, &payload)?;
+        self.invoke_command(
+            close_remote_tabs::COMMAND_NAME,
+            target,
+            &command_payload,
+            Some(close_remote_tabs::COMMAND_TTL),
+        )?;
+        Ok(())
+    }
+
+    pub(crate) fn handle_close_remote_tabs_command(
+        &mut self,
+        sender: Option<GetDeviceResponse>,
+        payload: serde_json::Value,
+    ) -> Result<IncomingDeviceCommand> {
+        let send_tab_key: PrivateSendTabKeys = match self.close_remote_tabs_key() {
+            Some(s) => PrivateSendTabKeys::deserialize(s)?,
+            None => {
+                return Err(Error::IllegalState(
+                    "Cannot find Close Remote Tabs keys. Has initialize_device been called before?",
+                ));
+            }
+        };
+        let encrypted_payload: EncryptedCloseTabsPayload = serde_json::from_value(payload)?;
+        match encrypted_payload.decrypt(&send_tab_key) {
+            Ok(payload) => Ok(IncomingDeviceCommand::TabsClosed { sender, payload }),
+            Err(e) => {
+                log::warn!("Could not decrypt Close Remote Tabs payload. Diagnosing then resetting the Close Tabs keys.");
+                self.clear_close_remote_tabs_keys();
+                self.reregister_current_capabilities()?;
+                Err(e)
+            }
+        }
+    }
+
+    pub(crate) fn load_or_generate_close_remote_tabs_keys(&mut self) -> Result<PrivateSendTabKeys> {
+        if let Some(s) = self.close_remote_tabs_key() {
+            match PrivateSendTabKeys::deserialize(s) {
+                Ok(keys) => return Ok(keys),
+                Err(_) => {
+                    error_support::report_error!(
+                        "fxaclient-close-tabs-key-deserialize",
+                        "Could not deserialize Close Remote Tabs keys. Re-creating them."
+                    );
+                }
+            }
+        }
+        let keys = PrivateSendTabKeys::from_random()?;
```

Comment:
(Yep, the type name is still `PrivateSendTab` keys, even though these are the "close tabs" keys, because it uses the same encryption scheme 😅 I'd be happy to rename these to something generic in a separate commit, or a follow-up!)

Reply:
A follow-up after the initial feature landed is definitely reasonable, otherwise I wonder if a comment here just mentioning we're using send tabs for convenience is good enough!